[CA-1084] [CA-1180] [CA-1213] RBS project permissions -- Remove resource buffer SA and add policy groups to custom IAM roles.

andy7i opened the pull request:
https://broadworkbench.atlassian.net/browse/CA-1084
https://broadworkbench.atlassian.net/browse/CA-1180

Most permissions are being set in RBS and at the folder level. This PR handles removing the RBS SA from the project as well as adding the policy groups to the appropriate new custom IAM roles. After this, we are no longer using DM for project creation.

The PRs related to this set of changes are:
https://github.com/broadinstitute/rawls/pull/1369
https://github.com/broadinstitute/terraform-terra/pull/144
https://github.com/broadinstitute/firecloud-develop/pull/2445
https://github.com/DataBiosphere/terra-resource-buffer/pull/120

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

andy7i committed:
Remove resource buffer SA and add policy groups to custom IAM roles.

andy7i committed:
test reference.conf fix

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -234,15 +240,31 @@ class HttpGoogleServicesDAO(
+
+      val policyGroupsToRoles = Map(
+        policyMap(SamWorkspacePolicyNames.owner) -> Set(googleProjectOwnerRole),
+        policyMap(SamWorkspacePolicyNames.canCompute) -> Set(googleProjectViewerRole)
```
using `policyMap` here because `policyGroupsByAccessLevel` seems intended to match google roles and does not include can-compute

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -234,15 +240,31 @@ class HttpGoogleServicesDAO(
+      )
+
+      for {
+        success <- policyGroupsToRoles.toList.traverse {case (email, roles) => googleIamDao.addIamRoles(GoogleProject(googleProject.value), email, MemberType.Group, roles)} // addIamRoles logic includes retries
```
I'm not sure if we also need to set up syncing the workspace policy groups to google similar to `syncBillingProjectOwnerPolicyToGoogleAndGetEmail`

andy7i committed:
The project-owner policy grants the (workspace) owner role, but the owner policy group does not include project-owners. Therefore, we need to explicitly add the project-owner here otherwise it wouldn't get the same custom iam role as an owner.

andy7i committed:
rename projectOwnerPolicyEmail to billingProjectOwnerPolicyEmail

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -234,15 +240,33 @@ class HttpGoogleServicesDAO(
     }
 
+    def updateGoogleProjectIam(googleProject: GoogleProjectId, policyMap: Map[SamResourcePolicyName, WorkbenchEmail], googleProjectOwnerRole: String, googleProjectViewerRole: String, billingProjectOwnerPolicyEmail: WorkbenchEmail): Future[List[Boolean]] = {
+      // billing project owner - organizations/$ORG_ID/roles/google-project-owner
```
Just want to double check that the billing project owner should get permissions here. The project-owner policy grants the (workspace) owner role, but the owner policy group does not include project-owners. That means that Sam sees that project-owners have the workspace owner permissions, but google does not. Therefore, we need to explicitly add the project-owner group here otherwise it wouldn't get the same custom iam role as the owner group.

andy7i committed:
Merge branch 'v2_workspaces' into aj--make-rbs-project-permissions-equivalent-to-dm--ca-1084--ca-1080

# Conflicts:
#	core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GoogleServicesDAO.scala
#	core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala
#	core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala
#	core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/MockGoogleServicesDAO.scala

andy7i committed:
related to resolving merge conflicts -- changed my mind mid-merge

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala:
```diff
@@ -134,6 +134,15 @@ object Boot extends IOApp with LazyLogging {
+    val pathToCredentialJson = gcsConfig.getString("pathToCredentialJson")
+    val source = scala.io.Source.fromFile(pathToCredentialJson)
+    val jsonCreds = try source.mkString finally source.close()
+    val googleIamDao = new HttpGoogleIamDAO(appName, GoogleCredentialModes.Json(jsonCreds), metricsPrefix)
```
GoogleIamDao should be part of AppDependencies and the code above (some of which is already there) should be in  initAppDependencies

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala:
```diff
@@ -134,6 +134,15 @@ object Boot extends IOApp with LazyLogging {
+    val jsonCreds = try source.mkString finally source.close()
+    val googleIamDao = new HttpGoogleIamDAO(appName, GoogleCredentialModes.Json(jsonCreds), metricsPrefix)
+    val googleProjectOwnerRole = gcsConfig.getString("googleProjectOwnerRole")
+    val googleProjectViewerRole = gcsConfig.getString("googleProjectViewerRole")
```
inline these 2 lines

dvoet reviewed (changes requested):
other comments in person

andy7i committed:
PR feedback in Boot.scala, moving httpGoogleIamDAO setup into initAppDependencies

andy7i committed:
refactor, pt 1 (doug suggested removing the nested traces since they don't provide much value but detract from readability ; moved the lines into a single for comp.)

andy7i committed:
update updateGoogleProjectIam permissions to grant billing project owners 2 roles and workspace writers 1 role. (to be renamed in a future commit)

andy7i committed:
refactor part 2: break up saveNewWorkspace into multiple Sam, Rawls parts: createPolicyMap, createWorkflowCollectionForWorkspace, syncPolicies

side note:

opted for
```
      for {
        _ <- traceDBIOWithParent("save", parentSpan)(_ => dataAccess.workspaceQuery.createOrUpdate(workspace))
      } yield(workspace)
```

instead of
```
      traceDBIOWithParent("save", parentSpan)(_ => dataAccess.workspaceQuery.createOrUpdate(workspace)).flatMap{_ => DBIO.successful(workspace)}
```
becase it's (a bit) easier to read

andy7i committed:
refactor part 3: move updateGoogleProjectIam from GoogleServicesDAO to setupGoogleProject in WorkspaceService

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -77,7 +81,8 @@ object WorkspaceService {
       maxActiveWorkflowsPerUser, workbenchMetricBaseName, submissionCostService,
-      config, requesterPaysSetupService, resourceBufferService)
+      config, requesterPaysSetupService, resourceBufferService, resourceBufferSaEmail,
+      googleIamDao: GoogleIamDAO, googleProjectOwnerRole: String, googleProjectViewerRole: String)
```
I didn't know you could put type definitions on params like this at call time.  Did you do this for a particular reason?  If not, I would prefer the data types be removed for consistency sake and so that if somebody else sees this in the future, they don't wonder if it was done for a particular reason.

andy7i replied:
whoops, that's from copy-pasting

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1902,16 +1907,32 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
     * @return Future[(GoogleProjectId, GoogleProjectNumber)] of the project that we claimed from RBS
     */
-  private def setupGoogleProject(billingProject: RawlsBillingProject, billingAccount: RawlsBillingAccountName, span: Span = null, workspaceId: String): Future[(GoogleProjectId, GoogleProjectNumber)] = {
+  private def setupGoogleProject(billingProject: RawlsBillingProject, billingAccount: RawlsBillingAccountName, span: Span = null, workspaceId: String, policyMap: Map[SamResourcePolicyName, WorkbenchEmail], billingProjectOwnerPolicyEmail: WorkbenchEmail): Future[(GoogleProjectId, GoogleProjectNumber)] = {
```
Update scaladoc too please.

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -2063,7 +2084,56 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+        }
+
+        // policyMap has policyName -> policyEmail
+        policyMap: Map[SamResourcePolicyName, WorkbenchEmail] = resource.accessPolicies.map(x => SamResourcePolicyName(x.id.accessPolicyName) -> WorkbenchEmail(x.email)).toMap
```
I was surprised to see the call to `samDAO.createResourceFull` in this method.  That call seems to me to be the most important thing we're doing here.  I sort of see 3 logical steps going on:

1. Define the the Sam Policies we want added to the Workspace when it is created in Sam
2. Create the Workspace resource in Sam
3. Make a `policyMap` that can be used by other methods

andy7i replied:
I think we worked out some of this together and it should be reflected in the latest code. We renamed the method to `createWorkspaceResourceInSam` and moved the policyMap out to withNewWorkspaceContext. (Maybe this comment was on an older commit or the github page hadn't updated? Please let me know if things still don't look right!)

andy7i committed:
unnested functions, bit more refactoring with Greg ; making sure futures run in parallel

andy7i committed:
Merge branch 'v2_workspaces' into aj--make-rbs-project-permissions-equivalent-to-dm--ca-1084--ca-1080

# Conflicts:
#	core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala
#	core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala
#	core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala
#	core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/ApiServiceSpec.scala
#	core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala

andy7i committed:
renaming custom iam role

andy7i committed:
the can-compute policy group doesn't actually contain the workspace owner, so I think it needs to be added manually

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1992,93 +2026,98 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+    for {
+      _ <- traceDBIOWithParent("save", parentSpan)(_ => dataAccess.workspaceQuery.createOrUpdate(workspace))
+    } yield(workspace)
+  }
```
it's a one-line for comp, but I preferred the readability of this compared to:

```
traceDBIOWithParent("save", parentSpan)(_ => dataAccess.workspaceQuery.createOrUpdate(workspace)).flatMap{_ => DBIO.successful(workspace)}
```

dvoet replied:
what about
```
  traceDBIOWithParent("save", parentSpan)(_ => dataAccess.workspaceQuery.createOrUpdate(workspace))
    .map(_ => workspace)
```

andy7i committed:
rename writer -> can-compute

andy7i committed:
this comment becomes redundant redundant

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1913,9 +1926,30 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+  private def updateGoogleProjectIam(googleProject: GoogleProjectId, policyMap: Map[SamResourcePolicyName, WorkbenchEmail], terraBillingProjectOwnerRole: String, terraWorkspaceCanComputeRole: String, billingProjectOwnerPolicyEmail: WorkbenchEmail): Future[List[Boolean]] = {
+    // organizations/$ORG_ID/roles/terra-billing-project-owner AND organizations/$ORG_ID/roles/terra-workspace-can-compute
+      // billing project owner
+    // organizations/$ORG_ID/roles/terra-workspace-can-compute
+      // workspace owner
+      // workspace can-compute
```
the can-compute policy group doesn't actually contain workspace owners, so I think it needs to be added separately.. I think this is an artifact of having hierarchical permissions being implemented in 2 ways: with groups (the "rawls way") and with descendant roles (the "sam way")

dvoet replied:
This is correct but not for the reason you describe. The reason is that the are 2 separate policies and policies generally have no relation to each other.

andy7i committed:
tracing for cloneWorkspace

andy7i committed:
fix intellij indentation-fail

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiService.scala:
```diff
@@ -83,8 +83,10 @@ trait WorkspaceApiService extends UserInfoDirectives {
             addLocationHeader(destWorkspace.toWorkspaceName.path) {
-              complete {
-                workspaceServiceConstructor(userInfo).CloneWorkspace(WorkspaceName(sourceNamespace, sourceWorkspace), destWorkspace).map(w => StatusCodes.Created -> WorkspaceDetails(w, destWorkspace.authorizationDomain.getOrElse(Set.empty)))
+              traceRequest { span =>
+                complete {
+                  workspaceServiceConstructor(userInfo).CloneWorkspace(WorkspaceName(sourceNamespace, sourceWorkspace), destWorkspace, span).map(w => StatusCodes.Created -> WorkspaceDetails(w, destWorkspace.authorizationDomain.getOrElse(Set.empty)))
+                }
```
cherry-picked clone workspace tracing from this PR to dev: https://github.com/broadinstitute/rawls/pull/1374
to eliminate a NPE

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -130,7 +135,7 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
   def GetTags(query: Option[String]) = getTags(query)
   def AdminListWorkspacesWithAttribute(attributeName: AttributeName, attributeValue: AttributeValue) = asFCAdmin { listWorkspacesWithAttribute(attributeName, attributeValue) }
-  def CloneWorkspace(sourceWorkspace: WorkspaceName, destWorkspace: WorkspaceRequest) = cloneWorkspace(sourceWorkspace, destWorkspace)
+  def CloneWorkspace(sourceWorkspace: WorkspaceName, destWorkspace: WorkspaceRequest, parentSpan: Span = null) = cloneWorkspace(sourceWorkspace, destWorkspace, parentSpan)
```
I don't think we want to change this line. The caller of `CloneWorkspace` should be the api routes layer which does not supply a span. Anywhere else should call `cloneWorkspace` directly.

andy7i replied:
CloneWorkspace is called within a `traceRequest` at the api routes layer (which is also how CreateWorkspace, GetWorkspace, and ListWorkspaces are called)

andy7i committed:
make parentSpan no longer optional, since the following line would throw an NPE:
```
            parentSpan.putAttribute("workspaceId", OpenCensusAttributeValue.stringAttributeValue(workspaceId))
```

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1913,9 +1927,30 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+
+    for {
+      success <- policyGroupsToRoles.toList.traverse {case (email, roles) => googleIamDao.addIamRoles(GoogleProject(googleProject.value), email, MemberType.Group, roles)} // addIamRoles logic includes retries
+    } yield (success)
```
Don't need to for comp here. Also put in a JIRA to do this in 1 call to `googleIamDao`. The issue is that each call to `googleIamDao.addIamRoles` is 2 google api calls assuming no retries, 1 to get and 1 to update. So as written we have 6 round trips to google. Google apis allow this can all be done in 1 pair of round trips but that requires adding a new function to `googleIamDao` to take the whole map.

andy7i replied:
created https://broadworkbench.atlassian.net/browse/CA-1220

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1992,93 +2027,98 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+        traceWithParent("createResourceFull (workspace)", parentSpan)(_ => samDAO.createResourceFull(SamResourceTypeNames.workspace, workspaceId, defaultPolicies, workspaceRequest.authorizationDomain.getOrElse(Set.empty).map(_.membersGroupName.value), userInfo, None))
+      }
+    } yield (resource)
+    )
```
for comp not required

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1992,93 +2027,98 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
-            val canCatalogPolicy = SamWorkspacePolicyNames.canCatalog -> SamPolicy(Set.empty, Set.empty, Set(SamWorkspaceRoles.canCatalog))
-
-            val defaultPolicies = Map(projectOwnerPolicy, ownerPolicy, writerPolicy, readerPolicy, shareReaderPolicy, shareWriterPolicy, canComputePolicy, canCatalogPolicy)
+    val defaultPolicies = Map(projectOwnerPolicy, ownerPolicy, writerPolicy, readerPolicy, shareReaderPolicy, shareWriterPolicy, canComputePolicy, canCatalogPolicy)
```
rename to `allPolicies`

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1992,93 +2027,98 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+  private def syncPolicies(workspaceId: String, policyMap: Map[SamResourcePolicyName, WorkbenchEmail], workspaceRequest: WorkspaceRequest, parentSpan: Span) = {
+    traceWithParent("traversePolicies", parentSpan) (s1 =>
+      Future.traverse(policyMap) { x =>
+        val policyName = x._1
```
```Future.traverse(policyMap.keys) { policyName =>```

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -2090,36 +2130,49 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+              billingProjectOwnerPolicyEmail <- traceDBIOWithParent("getPolicySyncStatus", parentSpan)(_ => DBIO.from(
+                  samDAO.getPolicySyncStatus(SamResourceTypeNames.billingProject, workspaceRequest.namespace, SamBillingProjectPolicyNames.owner, userInfo).map(_.email)))
+              resource <- createWorkspaceResourceInSam(workspaceId, billingProjectOwnerPolicyEmail, workspaceRequest, parentSpan)
+              // policyMap has policyName -> policyEmail
+              policyMap: Map[SamResourcePolicyName, WorkbenchEmail] = resource.accessPolicies.map(x => SamResourcePolicyName(x.id.accessPolicyName) -> WorkbenchEmail(x.email)).toMap
```
rename `policyMap` to `policyEmailsByName` and remove the comment

andy7i committed:
addressing PR feedback, updating previously existing code

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -2090,36 +2125,48 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+                } yield()})
+              // After the workspace has been created, create the google-project resource in Sam with the workspace as the resource parent
+              _ <- traceDBIOWithParent("createResourceFull (google project)", parentSpan)(_ => DBIO.from(
+                  samDAO.createResourceFull(SamResourceTypeNames.googleProject, googleProjectId.value, Map.empty, workspaceRequest.authorizationDomain.getOrElse(Set.empty).map(_.membersGroupName.value), userInfo, Option(SamFullyQualifiedResourceId(workspaceId, SamResourceTypeNames.workspace.value)))))
```
Could this call be added to the `for` comp just above so that this step gets run in parallel with two previous futures?

andy7i replied:
oh good point! I wonder if the following future (for `gcsDAO.setupWorkspace`) could also be run in parallel

andy7i replied:
I added gcsDAO.setupWorkspace to be run in parallel as well, which seems ok as far as I can tell

andy7i replied:
I asked doug to review and he had concerns about it being a premature optimization and suggested not parallelizing it until we have more data to support that the parallelization is worth it, so the last 2 commits revert the parallelization

andy7i committed:
PR feedback

andy7i committed:
run gcsDAO.setupWorkspace in parallel as well

andy7i committed:
Revert "run gcsDAO.setupWorkspace in parallel as well"

This reverts commit 15837587

andy7i committed:
Revert parallelization
